### ISSUE TITLE
is lightGBM custom eval metric usage possible?

### ISSUE THREAD

**Issue by flamby** [optuna/optuna#861]
Hi,

Is a custom eval metric function usable with Optuna?

As the eval metric argument of `LightGBMPruningCallback` type is a string, I can't make it work.

Thanks and keep the good work!

**Comment by crcrpar** [comment]
Hi @flamby,

Feel sorry to bother you and my delayed response, but could you share a minimal reproducible example?

**Comment by smly** [comment]
Hi @flamby,

Thanks for the good question!
I wrote an example code to use a custom metric function with Optuna.
https://gist.github.com/smly/5a5ddf968d59492b79e4cbf90b2d3430

Hope this code helps to solve your problem.

**Comment by flamby** [comment]
Hi @crcrpar and @smly 

Sorry for the late reply.

@smly gist patch is exactly what I was not able to achieve. Silly me, I used the very same numpy method for both the `feval` part and the `LightGBMPruningCallback` part.

In your patch, you did not choose to change `param["metric"]` to `"custom_accuracy_numpy"`. 
Is it because it's useless as it's overriden somewhere automatically by the custom function?

@crcrpar : A related question regarding lightgbm integration future : As I'm more used to the sklearn wrapper of lightgbm, I tend to prefer the [OptGBM](https://github.com/Y-oHr-N/OptGBM) syntax. Do you plan to merge it? I find it very convenient.

Thanks again

**Comment by smly** [comment]
> Is it because it's useless as it's overriden somewhere automatically by the custom function?

Partially correct. The evaluation result of "auc" is not used in my example code, but it doesn’t mean something is overridden. You can omit `param['metric']`.

More precisely, as described in [LightGBM document](https://lightgbm.readthedocs.io/en/latest/Parameters.html#metric), `param['metric']` is **the metric(s) to be evaluated on the evaluation set(s)**. It doesn't mean that `param['metric']` is used for pruning. The metric used for pruning is specified in the argument of `LightGBMPruningCallback`. In my example code, two evaluation metrics, "auc" and "custom_accuray", are evaluated. Then, "custom_accuracy" is used for pruning.

> As I'm more used to the sklearn wrapper of lightgbm, I tend to prefer the OptGBM syntax. Do you plan to merge it? I find it very convenient.

Thanks for the feedback! IMO, merging with OptGBM is a good option for Optuna, but it's necessary to discuss the detail with the author of OptGBM. I might discuss this topic with him in a one-day Optuna Development Sprint event on Mar 1.

**Comment by flamby** [comment]
Thanks for the clarification @smly 

and keep the good work!

closing then.

**Comment by mirekphd** [comment]
> I wrote an example code to use a custom metric function with Optuna.
> https://gist.github.com/smly/5a5ddf968d59492b79e4cbf90b2d3430
> 
> Hope this code helps to solve your problem.

It's important to note that second parameter of `LightGBMPruningCallback` accepts not the custom metric function name as intuition would suggest (here: `custom_accuracy_pct`), but the *custom metric name* (here: `"custom_accuracy"`) returned by this function:
![image](https://user-images.githubusercontent.com/36706320/131477152-6e1da57b-1aeb-4aba-84ef-558ca1d145f8.png)

Note also that custom metric name has to be hard-coded - defined as fixed string inside of the said function, because the function cannot be parametrized with an extra parameter... requiring separate function for each custom metric.
